File: bot/skills/mute.py
import logging
from datetime import datetime, timedelta
from random import choice
from typing import List

from telegram import Update, User, ChatPermissions
from telegram.error import TelegramError
from telegram.ext import ContextTypes
from mode import cleanup_queue_update
from handlers import ChatCommandHandler
from utils.time import get_duration
from typing_utils import App, get_job_queue

logger = logging.getLogger(__name__)

MIN_MUTE_TIME = timedelta(minutes=1)
MAX_MUTE_TIME = timedelta(days=365)


def add_mute(app: App, handlers_group: int):
    logger.info("registering mute handlers")
    app.add_handler(
        ChatCommandHandler(
            "mute",
            mute,
            require_admin=True,
        ),
        group=handlers_group,
    )
    app.add_handler(ChatCommandHandler("mute", mute_self), group=handlers_group)
    app.add_handler(
        ChatCommandHandler(
            "unmute",
            unmute,
            require_admin=True,
        ),
        group=handlers_group,
    )


def _get_minutes(args: List[str]) -> timedelta:
    # cmd should be a reply for going to mute user message like "/mute 90"
    if len(args) < 1:
        raise Exception(
            "mute cmd should be a reply for going to mute user message like '/mute 90', "
            "where '90' is duration of the mute"
        )
    return get_duration(args[0])


async def mute_user_for_time(
    update: Update,
    context: ContextTypes.DEFAULT_TYPE,
    user: User,
    mute_duration: timedelta,
):
    if update.message is None or update.effective_chat is None:
        return
    message = update.message
    mute_duration = max(mute_duration, MIN_MUTE_TIME)
    mute_duration = min(mute_duration, MAX_MUTE_TIME)
    try:
        until = datetime.now() + mute_duration
        logger.info(
            "user: %s[%d] will be muted for %s", user.full_name, user.id, mute_duration
        )

        await message.reply_text(f"Таймаут для {user.full_name} на {mute_duration}")
        mute_perm = ChatPermissions(
            can_add_web_page_previews=False,
            can_send_other_messages=False,
            can_send_messages=False,
            can_send_polls=False,
            can_send_audios=False,
            can_send_documents=False,
            can_send_photos=False,
            can_send_videos=False,
            can_send_video_notes=False,
            can_send_voice_notes=False,
        )
        await context.bot.restrict_chat_member(
            update.effective_chat.id, user.id, mute_perm, until
        )
    except TelegramError as err:
        logger.error("can't mute user %s: %s", user, err)
        await message.reply_text(f"😿 не вышло, потому что: \n\n{err}")


async def mute(update: Update, context: ContextTypes.DEFAULT_TYPE):
    if update.message is None or update.message.reply_to_message is None:
        return
    user: User | None = update.message.reply_to_message.from_user
    if user is None:
        return
    args = context.args or []
    mute_minutes = _get_minutes(args)
    await mute_user_for_time(update, context, user, mute_minutes)


async def mute_self(update: Update, context: ContextTypes.DEFAULT_TYPE):
    user: User | None = update.effective_user
    if user is None:
        return
    if update.message is None:
        return
    message = update.message
    await mute_user_for_time(update, context, user, timedelta(days=1))
    self_mute_messages = [
        f"Да как эта штука работает вообще, {user.name}?",
        f"Не озоруй, {user.name}, мало ли кто увидит",
        f"Зловив {user.name} на вила!",
        f"Насилие порождает насилие, {user.name}",
        f"Опять ты, {user.name}!",
    ]
    result = await message.reply_text(choice(self_mute_messages))

    cleanup_queue_update(
        get_job_queue(context),
        message,
        result,
        600,
        remove_cmd=True,
        remove_reply=False,
    )


async def unmute_user(
    update: Update, context: ContextTypes.DEFAULT_TYPE, user: User
) -> None:
    if update.message is None or update.effective_chat is None:
        return
    message = update.message
    try:
        await message.reply_text(f"{user.full_name}, не озоруй! Мало ли кто увидит 🧐")
        unmute_perm = ChatPermissions(
            can_add_web_page_previews=True,
            can_send_other_messages=True,
            can_send_messages=True,
            can_send_polls=True,
            can_send_audios=True,
            can_send_documents=True,
            can_send_photos=True,
            can_send_videos=True,
            can_send_video_notes=True,
            can_send_voice_notes=True,
            can_invite_users=True,
        )
        await context.bot.restrict_chat_member(
            update.effective_chat.id, user.id, unmute_perm
        )
    except TelegramError as err:
        await message.reply_text(f"😿 не вышло, потому что: \n\n{err}")


async def unmute(update: Update, context: ContextTypes.DEFAULT_TYPE) -> None:
    if update.message is None or update.message.reply_to_message is None:
        return
    user: User | None = update.message.reply_to_message.from_user
    if user is None:
        return
    await unmute_user(update, context, user)

File: internal/skill/mute.go
package skill

import (
	"context"
	"fmt"
	"log/slog"
	"math/rand/v2"
	"time"

	"github.com/go-telegram/bot"
	"github.com/go-telegram/bot/models"

	appbot "github.com/vldc-hq/vldc-bot/internal/bot"
	"github.com/vldc-hq/vldc-bot/internal/util"
)

const defaultCleanupDelay = 600 * time.Second

const (
	minMuteTime = time.Minute
	maxMuteTime = 365 * 24 * time.Hour
)

var selfMuteMessages = []string{
	"Да как эта штука работает вообще, %s?",
	"Не озоруй, %s, мало ли кто увидит",
	"Зловив %s на вила!",
	"Насилие порождает насилие, %s",
	"Опять ты, %s!",
}

func MuteSkill() appbot.Skill {
	return appbot.Skill{
		Name: "mute",
		Hint: "mute/unmute users",
		Register: func(b *bot.Bot, _ *appbot.Deps) {
			b.RegisterHandler(bot.HandlerTypeMessageText, "/mute", bot.MatchTypePrefix, handleMute)
			b.RegisterHandler(bot.HandlerTypeMessageText, "/unmute", bot.MatchTypePrefix, handleUnmute)
		},
	}
}

func handleMute(ctx context.Context, b *bot.Bot, update *models.Update) {
	if update.Message == nil {
		return
	}
	msg := update.Message

	// Self-mute: /mute without reply
	if msg.ReplyToMessage == nil {
		user := msg.From
		if user == nil {
			return
		}
		MuteUserForDuration(ctx, b, msg.Chat.ID, user.ID, user.FirstName, 24*time.Hour)
		text := fmt.Sprintf(selfMuteMessages[rand.IntN(len(selfMuteMessages))], displayName(user))
		result, _ := b.SendMessage(ctx, &bot.SendMessageParams{
			ChatID:          msg.Chat.ID,
			Text:            text,
			ReplyParameters: &models.ReplyParameters{MessageID: msg.ID},
		})
		util.ScheduleCleanup(b, msg.Chat.ID, defaultCleanupDelay, msg.ID, msgID(result))
		return
	}

	// Admin mute: /mute [duration] as reply
	if !appbot.IsAdmin(ctx, b, msg.Chat.ID, msg.From.ID) {
		return
	}

	target := msg.ReplyToMessage.From
	if target == nil {
		return
	}

	args := extractArgs(msg.Text, "/mute")
	if args == "" {
		_, _ = b.SendMessage(ctx, &bot.SendMessageParams{
			ChatID:          msg.Chat.ID,
			Text:            "Usage: /mute <duration> (e.g., /mute 90, /mute 2h30m)",
			ReplyParameters: &models.ReplyParameters{MessageID: msg.ID},
		})
		return
	}

	duration := util.ParseDuration(args)
	if duration <= 0 {
		duration = time.Minute
	}

	MuteUserForDuration(ctx, b, msg.Chat.ID, target.ID, target.FirstName, duration)
}

func handleUnmute(ctx context.Context, b *bot.Bot, update *models.Update) {
	if update.Message == nil || update.Message.ReplyToMessage == nil {
		return
	}
	msg := update.Message
	if !appbot.IsAdmin(ctx, b, msg.Chat.ID, msg.From.ID) {
		return
	}

	target := msg.ReplyToMessage.From
	if target == nil {
		return
	}

	UnmuteUser(ctx, b, msg.Chat.ID, target.ID, target.FirstName)
	_, _ = b.SendMessage(ctx, &bot.SendMessageParams{
		ChatID:          msg.Chat.ID,
		Text:            fmt.Sprintf("%s, не озоруй! Мало ли кто увидит 🧐", target.FirstName),
		ReplyParameters: &models.ReplyParameters{MessageID: msg.ID},
	})
}

// MuteUserForDuration restricts a user in the chat for the given duration.
func MuteUserForDuration(ctx context.Context, b *bot.Bot, chatID, userID int64, userName string, duration time.Duration) {
	duration = max(duration, minMuteTime)
	duration = min(duration, maxMuteTime)
	until := time.Now().Add(duration)

	slog.Info("muting user", "user", userName, "user_id", userID, "duration", duration)

	_, err := b.RestrictChatMember(ctx, &bot.RestrictChatMemberParams{
		ChatID:    chatID,
		UserID:    userID,
		UntilDate: int(until.Unix()),
		Permissions: &models.ChatPermissions{
			CanSendMessages:       false,
			CanSendAudios:         false,
			CanSendDocuments:      false,
			CanSendPhotos:         false,
			CanSendVideos:         false,
			CanSendVideoNotes:     false,
			CanSendVoiceNotes:     false,
			CanSendPolls:          false,
			CanSendOtherMessages:  false,
			CanAddWebPagePreviews: false,
		},
	})
	if err != nil {
		slog.Error("failed to mute user", "user", userName, "error", err)
		_, _ = b.SendMessage(ctx, &bot.SendMessageParams{
			ChatID: chatID,
			Text:   fmt.Sprintf("😿 не вышло, потому что: \n\n%s", err),
		})
		return
	}

	_, _ = b.SendMessage(ctx, &bot.SendMessageParams{
		ChatID: chatID,
		Text:   fmt.Sprintf("Таймаут для %s на %s", userName, duration),
	})
}

// UnmuteUser restores all permissions for a user.
func UnmuteUser(ctx context.Context, b *bot.Bot, chatID, userID int64, userName string) {
	slog.Info("unmuting user", "user", userName, "user_id", userID)

	_, err := b.RestrictChatMember(ctx, &bot.RestrictChatMemberParams{
		ChatID: chatID,
		UserID: userID,
		Permissions: &models.ChatPermissions{
			CanSendMessages:       true,
			CanSendAudios:         true,
			CanSendDocuments:      true,
			CanSendPhotos:         true,
			CanSendVideos:         true,
			CanSendVideoNotes:     true,
			CanSendVoiceNotes:     true,
			CanSendPolls:          true,
			CanSendOtherMessages:  true,
			CanAddWebPagePreviews: true,
			CanInviteUsers:        true,
		},
	})
	if err != nil {
		slog.Error("failed to unmute user", "user", userName, "error", err)
	}
}

func msgID(m *models.Message) int {
	if m == nil {
		return 0
	}
	return m.ID
}

func displayName(user *models.User) string {
	if user.Username != "" {
		return "@" + user.Username
	}
	return user.FirstName
}

func extractArgs(text, command string) string {
	if len(text) <= len(command) {
		return ""
	}
	rest := text[len(command):]
	if rest != "" && rest[0] == '@' {
		// strip @botname
		idx := 0
		for idx < len(rest) && rest[idx] != ' ' {
			idx++
		}
		rest = rest[idx:]
	}
	if rest != "" && rest[0] == ' ' {
		rest = rest[1:]
	}
	return rest
}
